### PR TITLE
[Client] Base structure for client app, with React+Redux+Tests (and some typescript)

### PR DESCRIPTION
**Pull request by JabX**
*Before I forget, it looks like npm reformats the package.json with 2-space indentation every time it modifies it, so please stop trying to force 4-space indentation on that, thanks.*

There's a lot to cover with this PR.

## Code organisation
As we've seen before, every common code between the two apps resides under the `common` folder, and only the specific stuff has be to in the app folder. Basically, all that should reside there are the views (which should only use components from common), the routing, and probably some styling. That means there's no direct use of Redux stuff in there, to keep concerns separated.

Under common, Redux's `actions`, `reducers` and `store` are located under their respective folders, which are completed by a self-explanatory `components` folder, a `server` folder which contains the api definition, implementation and mock, and a `definitions` folder which contains shared Typescript definitions among all those folders.

## Redux
Every state node should have his own reducer in his own file and his own actions in their own file. Each node should also have TS state and action definitions. For instance, the `articleList` state node is specified by the reducer in `reducers/article-list.ts`, the actions in `actions/article-list.ts` and the definitions in `definitions/article-list.d.ts`.

### Actions
All possible actions are defined as a Typescript enum in `actions/index.ts`. Only the "macro" action creators should be exported and they should be thunks that handle the complete process. For instance, `articleList` has one exported action creator, `loadArticleList`, which dispatches `RequestArticleList', calls the API, and then dispatches either `ReceiveArticleList' or `FailureArticleList`.

#### Testing
There's a very convenient library called `redux-actions-assertions` which extends `chai` with methods to test that thunk actions correctly dispatch any given actions. So we only need to test the "exported" actions for the action they dispatch.

### Reducers
Each reducer should define an initial state and also export it.

#### Testing
I believe that since our reducers are mostly straightforward functions, the only test we should commit to write is the `rootReducer` test. In that test, the store state is initialized with all default values from reducers and we test each action one by one.

## API and mock
We can initialize the thunk middleware with an object to pass around to each thunk action creators, so we'll use this to pass around an API object that contains functions to call the server. In the app, it's initialized with the real object with real calls to the server, and for testing it's initialized with a mock object that returns promises with fake data.

There's an API object interface in `server/index.d.ts` that describe the API object. When adding a new server call, it should be updated first. That way, we can ensure easily that the real object and the mock stay always up-to-date (well, that's how proper dependency injection works).

### CODE REVIEW

File: app/src/common/components/article/line.jsx [hunk 1]
```diff
@@ -0,0 +1,4 @@
+/** Component that displays an article as a line. */
+export function ArticleLine({article}) {
+    return <li>{article.title} {article.description} {article.content}</li>;
```
Comment (pierr) [inline app/src/common/components/article/line.jsx]: Add `ArticleLine.displayName = 'ArticleLine'``

File: app/src/common/components/article/list.jsx [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import {connect} from 'react-redux';
+import {Component, PropTypes} from 'react';
+
+import {loadArticleList} from '../../actions/article-list';
+import {ArticleLine} from './line';
+
+/** Component that displays the list of all articles, connected to the store. */
+@connect(
```
Comment (pierr) [inline app/src/common/components/article/list.jsx]: Can you user the `compose` function of redux and not eh annotation. And the function enhancement.

File: app/src/common/reducers/article-list.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import {Action} from '../actions/index';
+import {ArticleListState, ArticleListAction} from '../definitions/article-list';
+
+export const defaultValue: ArticleListState = {isLoading: false, list: []};
+
+/** Reducer that handles the `articleList` store node. */
+export function articleList(state: ArticleListState = defaultValue, action: ArticleListAction) {
+    switch (action.type) {
+        case Action.RequestArticleList:
```
Comment (pierr) [inline app/src/common/reducers/article-list.ts]: use `REQUEST_ARTICLE_LIST`

File: app/test/mocha-init.js [hunk 4]
```diff
@@ -1,38 +1,84 @@
...
+const babel = require('babel-core');
+
+const babelOpts = {
+    presets: [require('babel-preset-node6')],
+    ast: false
+};
+
+var tsLoader = require.extensions['.ts', '.tsx'];
+
+var property = {
```
Comment (pierr) [inline app/test/mocha-init.js]: let or const

File: app/src/common/components/article/line.jsx (hunk 1 above)
Comment (JabX) [inline app/src/common/components/article/line.jsx]: I was under the impression what the displayName wasn't necessary if the function or class already had a name. Anyway this is just a placeholder component that's going to be replaced very soon.

**Pull request merged**